In [1]:
import pandas as pd
import numpy as np

In [3]:
car_sales_missing = pd.read_csv("data/car-sales-extended-missing-data.csv")

In [4]:
car_sales_missing.head()

Make Colour  Odometer (KM)  Doors    Price
0   Honda  White        35431.0    4.0  15323.0
1     BMW   Blue       192714.0    5.0  19943.0
2   Honda  White        84714.0    4.0  28343.0
3  Toyota  White       154365.0    4.0  13434.0
4  Nissan   Blue       181577.0    3.0  14043.0

In [8]:
car_sales_missing.isna().count()


Make             1000
Colour           1000
Odometer (KM)    1000
Doors            1000
Price            1000
dtype: int64

In [9]:
x = car_sales_missing.drop("Price",axis=1)
y = car_sales_missing["Price"]

# Fill Columns

In [14]:
car_sales_missing["Make"].fillna("missing", inplace=True)
car_sales_missing["Colour"].fillna("missing", inplace=True)
car_sales_missing["Odometer (KM)"].fillna(car_sales_missing["Odometer (KM)"].mean(), inplace=True)
car_sales_missing["Doors"].fillna(4,inplace=True)

car_sales_missing.isna().sum()

Make              0
Colour            0
Odometer (KM)     0
Doors             0
Price            50
dtype: int64

# For price drop all

In [19]:
car_sales_missing.dropna(inplace=True)
car_sales_missing.isna().sum()


Make             0
Colour           0
Odometer (KM)    0
Doors            0
Price            0
dtype: int64

In [20]:
len(car_sales_missing)

950

In [21]:
x = car_sales_missing.drop("Price",axis=1)
y = car_sales_missing["Price"]

In [22]:
#converting our data
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make","Colour","Doors"]

one_hot = OneHotEncoder()
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)]
                               ,remainder="passthrough")

In [24]:
try:
    transformed_x = transformer.fit_transform(car_sales_missing)
    print("Transformed features:\n", transformed_x)
except Exception as e:
    print("An error occurred during fit_transform:", e)

Transformed features:
 [[0.00000e+00 1.00000e+00 0.00000e+00 ... 0.00000e+00 3.54310e+04
  1.53230e+04]
 [1.00000e+00 0.00000e+00 0.00000e+00 ... 1.00000e+00 1.92714e+05
  1.99430e+04]
 [0.00000e+00 1.00000e+00 0.00000e+00 ... 0.00000e+00 8.47140e+04
  2.83430e+04]
 ...
 [0.00000e+00 0.00000e+00 1.00000e+00 ... 0.00000e+00 6.66040e+04
  3.15700e+04]
 [0.00000e+00 1.00000e+00 0.00000e+00 ... 0.00000e+00 2.15883e+05
  4.00100e+03]
 [0.00000e+00 0.00000e+00 0.00000e+00 ... 0.00000e+00 2.48360e+05
  1.27320e+04]]


## Option 2
    Fill with sklearn

In [25]:
car_sales_missing = pd.read_csv("data/car-sales-extended-missing-data.csv")

In [27]:
car_sales_missing.head()


Make Colour  Odometer (KM)  Doors    Price
0   Honda  White        35431.0    4.0  15323.0
1     BMW   Blue       192714.0    5.0  19943.0
2   Honda  White        84714.0    4.0  28343.0
3  Toyota  White       154365.0    4.0  13434.0
4  Nissan   Blue       181577.0    3.0  14043.0

In [29]:
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

In [30]:
car_sales_missing.dropna(subset=["Price"], inplace=True)
car_sales_missing.isna().sum()

Make             47
Colour           46
Odometer (KM)    48
Doors            47
Price             0
dtype: int64

In [31]:
x = car_sales_missing.drop("Price", axis=1)
y = car_sales_missing["Price"]

In [34]:
# Fill mising value with Scikit-Learn
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer

# Fill categorical value with 'missing & numerical values with mean'
cat_imputer = SimpleImputer(strategy="constant",fill_value="missing") #for every catagorical value constantly fill "missing"
door_imputer = SimpleImputer(strategy="constant",fill_value=4)
num_imputer = SimpleImputer(strategy="mean")

#Define Columns
cat_features = ["Make", "Colour"]
door_features = ["Doors"]
num_features = ["Odometer (KM)"]

# Create an imputer (someting that fills missing data)
imputer = ColumnTransformer([
    ("cat_imputer", cat_imputer, cat_features),
    ("door_imputer", door_imputer, door_features),
    ("num_imputer", num_imputer, num_features)
])

# Transform the data
filled_x = imputer.fit_transform(x)
filled_x

array([['Honda', 'White', 4.0, 35431.0],
       ['BMW', 'Blue', 5.0, 192714.0],
       ['Honda', 'White', 4.0, 84714.0],
       ...,
       ['Nissan', 'Blue', 4.0, 66604.0],
       ['Honda', 'White', 4.0, 215883.0],
       ['Toyota', 'Blue', 4.0, 248360.0]], dtype=object)

In [40]:
car_sales_filled = pd.DataFrame(filled_x,
                              columns=["Make", "Colour", "Doors", "Odometer (KM)"] )

In [41]:
car_sales_filled.head(10)

Make Colour Doors  Odometer (KM)
0    Honda  White   4.0        35431.0
1      BMW   Blue   5.0       192714.0
2    Honda  White   4.0        84714.0
3   Toyota  White   4.0       154365.0
4   Nissan   Blue   3.0       181577.0
5    Honda    Red   4.0        42652.0
6   Toyota   Blue   4.0       163453.0
7    Honda  White   4.0  130987.447894
8  missing  White   4.0       130538.0
9    Honda   Blue   4.0        51029.0

In [42]:
car_sales_filled.isna().sum()


Make             0
Colour           0
Doors            0
Odometer (KM)    0
dtype: int64

In [43]:
#converting our data
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make","Colour","Doors"]

one_hot = OneHotEncoder()
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)]
                               ,remainder="passthrough")

try:
    transformed_x = transformer.fit_transform(car_sales_filled)
    print("Transformed features:\n", transformed_x)
except Exception as e:
    print("An error occurred during fit_transform:", e)

Transformed features:
   (0, 1)	1.0
  (0, 9)	1.0
  (0, 12)	1.0
  (0, 14)	35431.0
  (1, 0)	1.0
  (1, 6)	1.0
  (1, 13)	1.0
  (1, 14)	192714.0
  (2, 1)	1.0
  (2, 9)	1.0
  (2, 12)	1.0
  (2, 14)	84714.0
  (3, 3)	1.0
  (3, 9)	1.0
  (3, 12)	1.0
  (3, 14)	154365.0
  (4, 2)	1.0
  (4, 6)	1.0
  (4, 11)	1.0
  (4, 14)	181577.0
  (5, 1)	1.0
  (5, 8)	1.0
  (5, 12)	1.0
  (5, 14)	42652.0
  (6, 3)	1.0
  :	:
  (943, 14)	162523.0
  (944, 0)	1.0
  (944, 6)	1.0
  (944, 11)	1.0
  (944, 14)	163322.0
  (945, 3)	1.0
  (945, 5)	1.0
  (945, 12)	1.0
  (945, 14)	35820.0
  (946, 4)	1.0
  (946, 9)	1.0
  (946, 11)	1.0
  (946, 14)	155144.0
  (947, 2)	1.0
  (947, 6)	1.0
  (947, 12)	1.0
  (947, 14)	66604.0
  (948, 1)	1.0
  (948, 9)	1.0
  (948, 12)	1.0
  (948, 14)	215883.0
  (949, 3)	1.0
  (949, 6)	1.0
  (949, 12)	1.0
  (949, 14)	248360.0


In [45]:
# Now we've got data as numbers and filled

np.random.seed(42)
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(transformed_x,y,test_size=0.2)

model = RandomForestRegressor()
model.fit(x_train, y_train)
model.score(x_test, y_test)

0.21990196728583944

# This one performed worse cuz it has only 950 data to train with ie 950 vs 1000
## --x-- End of filling mising data . --x--